Create clues2nrfconnect.py

0xAbby opened the pull request:
A simple script to convert CLUES format to nrfc connect's json format

0xAbby committed:
Create clues2nrfconnect.py

A simple script to convert CLUES format to nrfc connect app

0xAbby committed:
Update clues2nrfconnect.py

XenoKovah commented:
Did you test this with nRF Connect and find that it works? Because it didn't seem to be working for me on iOS, and more generally the output format doesn't look like it conforms to the expected nRF Connect format

0xAbby committed:
Update clues2nrfconnect.py

latest nrf connect no longer support previously defined structure found here:
https://github.com/NordicSemiconductor/bluetooth-numbers-database

0xAbby commented:
Yes, it was based on structure defined here [1] for NRF connect, unfortunately it doesnt seem to be supported on latest nrf connect apps.
I've updated the script and tested it on latest nrf connect (android)
[1] https://github.com/NordicSemiconductor/bluetooth-numbers-database

XenoKovah commented:
Yeah that https://github.com/NordicSemiconductor/bluetooth-numbers-database isn't related to their custom UUID format for the app at all.

I think the code isn't correct still, because the uuid128bitCharacteristicDefinitions dictionary is only 3 entries, when in reality the majority of the CLUES data are Characteristic UUIDs (you'll need to iterate through the `UUID_usage_array` and look for ``GATT Characteristic`). And then likewise only things in the `UUID_usage_array` tagged as "GATT Service" should be in the `uuid16bitServiceDefinitions` or `uuid128bitServiceDefinitions` entries of the output.

And while you're making changes, in order to make things more semantically meaningful for users, let's instead of using use the UUID name as-is, let's have it be `f"{entry['company'} - {entry['UUID_name'}"` from the CLUES data (using "Unknown" wherever `UUID_name` isn't populated) as the final nRF Connect "name" field.

Also I'm not sure why the script is trying to find different names in pick_name() (other than that I assume it was AI coded? ;)). CLUES is a [*schema*](https://darkmentor.com/CLUES_Schema/CLUES.html), (that comes with some example data) so the UUID name field shall always be named exactly "UUID_name". But as the schema doesn't state it's required, that means it can also be absent, and then you should just use "Unknown" (because at a minimum there will be a company name, and it will interpolate to <company> - Unknown.

0xAbby committed:
Refactor pick_name for better output naming

Updated the pick_name function to provide a more user-friendly output name by incorporating company and UUID name information.

0xAbby commented:
Changes made to include 128uuid and add company name. Tested it on NRF Connect version 4.29.1 (Android).
Note that you might need to clear cache/storage of the NRF connect App, in order to correctly import the output converted definitions file and be able to test it.

Examples:
<img width="1080" height="2400" alt="NRFConnect_android2" src="https://github.com/user-attachments/assets/7ceef3e6-6923-4ce3-aa79-10e6e3117041" />
<img width="1080" height="2400" alt="NRFConnect_android1" src="https://github.com/user-attachments/assets/39421075-463b-4e60-bf68-b3b001c0bfdc" />

XenoKovah commented:
It looks to me like this needs further revisions. E.g. in your above example you only tested that a custom service showed up. I can confirm that works on my Tesla. However I also confirmed that the Characteristics (e.g. 00000212-b2d1-43f0-9b88-960cebf8b91e "To Vehicle") were _not_ showing up. Make sure you drill down on the UVP and confirm you see service data being pulled in properly and reported too.

0xAbby committed:
Refactor load_entries and update metadata

Updated the last modified date and refactored the load_entries function to improve clarity and efficiency. Added new helper functions for better UUID handling and type hinting.

0xAbby commented:
made changes to fix the issue for mixing up the UUID for services/characteristics.

0xAbby committed:
createing/updating cleus2nrfconnect converter script
